onnxruntime/core/optimizer/common_subexpression_elimination.cc

```diff
@@ -0,0 +1,428 @@
+    return std::hash<const NodeArg*>{}(Normalize(node_arg));
+  }
+};
+
+struct NodeArgPtrEquality {
+  bool operator()(const NodeArg* lhs, const NodeArg* rhs) const {
+    return Normalize(lhs) == Normalize(rhs);
+  }
+};
+
+bool IsNodeSupported(const Node& node) {
+  if (node.Domain() != kOnnxDomain || node.ContainsSubgraph()) {
+    return false;
+  }
+
+  if (kNonDeterministicOps.find(node.OpType()) != kNonDeterministicOps.end()) {
+    return false;
+  }
+
+  return true;
+}
+}
+
+}
+
+namespace std {
+template <>
+struct hash<onnxruntime::EquivalenceClass> {
+  std::size_t operator()(const onnxruntime::EquivalenceClass& val) const noexcept {
+    return val.hash_;
+  }
+};
+}
+
+namespace onnxruntime {
+
+Status CommonSubexpressionElimination::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  // Pool of equivalence classes; unique_ptr to guarantee stable address.
+  std::vector<std::unique_ptr<EquivalenceClass>> unique_equivalence_classes;
+
+  // Maps an equivalence class of values to a representative NodeArg that belongs to this class.
+  std::unordered_map<
+    const EquivalenceClass*,
+    Representative,
+    DeepPointerHash, DeepPointerEquality> value_to_representative;
+
+  // Reverse mapping.
+  std::unordered_map<const NodeArg*, const EquivalenceClass*, NodeArgPtrHash, NodeArgPtrEquality> equivalence_classes;
+
+  int unique_discriminator = 1;
+
+  for (NodeIndex node_index : node_topology_list) {
+    const Node* node = graph_viewer.GetNode(node_index);
+    if (node == nullptr)
+      continue;
+
+    ORT_RETURN_IF_ERROR(Recurse(*graph.GetNode(node_index), modified, graph_level, logger));
+
+    std::vector<const EquivalenceClass*> input_values;
+    input_values.reserve(node->InputDefs().size());
+    for (const NodeArg* input_def : node->InputDefs()) {
+      auto it = equivalence_classes.find(input_def);
+      if (it == equivalence_classes.end()) {
+        // Because nodes are processed in topological order, this will always be
+        // a non-op value (graph input or constant initializer).
+        auto value = onnxruntime::make_unique<EquivalenceClass>(input_def);
+        const auto* raw_ptr = value.get();
+        unique_equivalence_classes.emplace_back(std::move(value));
+        value_to_representative.emplace(raw_ptr, Representative{input_def, 0, kInvalidOutputIndex});
+        it = equivalence_classes.emplace_hint(it, input_def, raw_ptr);
+      }
+
+      input_values.push_back(it->second);
+    }
+
+    int discriminator = 0;
+    if (!IsNodeSupported(*node)) {
+      discriminator = ++unique_discriminator;
+    }
+
+    for (OutputIndex output_index = 0; output_index < static_cast<int>(node->OutputDefs().size()); ++output_index) {
+      const NodeArg* output_def = node->OutputDefs()[output_index];
```

Comment:
nit: prefer calling size() once if it's not going to change across iterations. Not a big deal here but a good general approach to take.

e.g.

for (OutputIndex output_index = 0, end = static_cast<OutputIndex>(node->OutputDefs().size()); output_index < end; ++output_index) {